src/tools/perf/perftest_daemon.c

```diff
@@ -96,6 +107,287 @@ ucp_perf_daemon_server_conn_handle_cb(ucp_conn_request_h conn_request,
+    ep_params.err_mode         = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.err_handler.cb   = ucp_perf_daemon_err_cb;
+    ep_params.err_handler.arg  = ctx;
+    ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+    ep_params.sockaddr.addr    = (struct sockaddr*)&ctx->peer_address;
+    ep_params.sockaddr.addrlen = sizeof(ctx->peer_address);
+
+    status = ucp_ep_create(ctx->worker, &ep_params, &ctx->peer_ep);
+    if (status != UCS_OK) {
+        ucs_fatal("daemon failed to create an endpoint: %s",
+                  ucs_status_string(status));
+    }
+
+    ucp_perf_daemon_send_am_eager_reply(ctx->peer_ep,
+                                        UCP_PERF_DAEMON_AM_ID_PEER_INIT);
+}
+
+static void ucp_perf_daemon_check_am_msg(const ucp_am_recv_param_t *param,
+                                         size_t header_length,
+                                         int check_reply_ep,
+                                         int check_rndv,
+                                         ucp_perf_daemon_am_id_t am_id)
+{
+    ucs_debug("message received: %s", ucp_perf_daemon_am_id_name(am_id));
+
+    if (check_rndv) {
+        if (!(param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV)) {
+            ucs_fatal("am message received with unsupported eager protocol");
+        }
+    } else {
+        ucs_assert(!(param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV));
+    }
+    ucs_assertv(header_length == 0, "header_length %zu", header_length);
+
+    if (check_reply_ep) {
+        ucs_assert(param->recv_attr & UCP_AM_RECV_ATTR_FIELD_REPLY_EP);
+    }
+}
+
+static ucp_mem_h
+ucp_perf_daemon_memh_import(ucp_perf_daemon_context_t *ctx, void *packed_memh)
+{
+    ucp_mem_map_params_t params = {};
+    ucs_status_t status;
+    ucp_mem_h memh;
+
+    params.field_mask           = UCP_MEM_MAP_PARAM_FIELD_EXPORTED_MEMH_BUFFER;
+    params.exported_memh_buffer = packed_memh;
+    status                      = ucp_mem_map(ctx->context, &params, &memh);
+    if (status != UCS_OK) {
+        ucs_fatal("failed to import memory (%s)", ucs_status_string(status));
+    }
+
+    return memh;
+}
+
+static void *
+ucp_perf_unpack_length_value(void *ptr, uint16_t *length, void **value)
+{
+    *length = *ucs_serialize_next(&ptr, uint16_t);
+    *value = (length != 0) ? ucs_serialize_next_raw(&ptr, void, *length) : NULL;
+    return ptr;
+}
+
+static ucs_status_t
+ucp_perf_daemon_host_init_handler(void *arg, const void *header,
+                                  size_t header_length, void *data,
+                                  size_t length,
+                                  const ucp_am_recv_param_t *param)
+{
+    ucp_perf_daemon_context_t *ctx  = arg;
+    void *ptr                       = data;
+    void *value;
+    uint16_t value_length;
+
+    ucp_perf_daemon_check_am_msg(param, header_length, 1, 0,
+                                 UCP_PERF_DAEMON_AM_ID_HOST_INIT);
+
+    if (ctx->peer_ep != NULL) {
```

Comment:
why checking `peer_ep` here?

Reply:
I borrowed this code from the initial implementation, and IMO it is correct (at least at this stage).

On startup host process connects to daemon and send HOST_INIT message to it, containing IP:port of the peer daemon and exported memory handles. So on the initial HOST_INIT the remote daemon peer is not yet known, and this is expected behavior.

What if `peer_ep` is already resolved? That could mean that this is a stale daemon process that was already configured once, and now we receive duplicate HOST_INIT message. Or it receives HOST_INIT from multiple host processes. IMO at this stage of the daemon we can just abort the process, this is what we do.

Maybe in the future we need to better handle that, any thoughts?
Like persistent daemon service, or multiple host processes supported

Reply:
ok, can you pls add a short comment here for this check?

Reply:
sure